---
title: "Coral_Nursery_Amplicons"
author: "J. Meyer"
date: "2025-05-01"
output: html_document
---

```{r, echo=FALSE}
library(dada2)
library(phyloseq)
library(CoDaSeq)
library(vegan)
library(pairwiseAdonis)
library(corncob)
library(ggplot2)
library(randomcoloR)
library(cowplot)
library(knitr)
library(dplyr)
library(reshape2)
library(tibble)
writeLines(capture.output(sessionInfo()), "sessionInfo.txt")
```

## Quality-filter the sequencing reads and create Amplicon Sequence Variant (ASV) table with DADA2

Put unjoined R1 and R2 fastq files, with adaptors and primers previously removed with cutadapt into a directory for DADA2. Here, our forward and reverse fastq filenames have format: SAMPLENAME_R1_cut.fastq.gz and SAMPLENAME_R2_cut.fastq.gz


```{r, echo=FALSE}
path <- "~/Documents/Coral_Nursery_CNAT_DLAB/Coral_Nursery_cutadapt"
list.files(path)
fnFs <- sort(list.files(path, pattern="_R1_cut.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_cut.fastq.gz", full.names = TRUE))
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
# Perform filtering and trimming
filt_path <- file.path(path, "filtered") 
filtFs <- file.path(filt_path, paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(filt_path, paste0(sample.names, "_R_filt.fastq.gz"))
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(150,150),
                     maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
                     compress=TRUE, multithread=TRUE)
head(out)
# Learn the Error Rates, it TAKES TIME!
errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)
plotErrors(errF, nominalQ=TRUE)
# Dereplicate the filtered fastq files
derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)
names(derepFs) <- sample.names
names(derepRs) <- sample.names
# Infer the sequence variants in each sample
dadaFs <- dada(derepFs, err=errF, multithread=TRUE)
dadaRs <- dada(derepRs, err=errR, multithread=TRUE)
# Inspecting the dada-class object returned by dada:
dadaFs[[1]]
# Merge the denoised forward and reverse reads:
mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])
# Construct sequence table
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
#Remove chimeric sequences:
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab) # proportion of non-chimeric taxa
#Track reads through the pipeline
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoised", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
write.table(track, "dada_read_stats.txt",sep="\t",col.names=NA)
saveRDS(seqtab.nochim, "~/Documents/Coral_Nursery_CNAT_DLAB/seqtab.nochim.rds")
```


## Data quality check
I inspected the dada_read_stats.txt file and have 18 biological samples that had fewer than 500 reads after quality filtering, denoising, merging, and removal of chimeras. I will merge the dada_read_stats file add the number of non-chimeric reads to the metadata table for downstream filtering of those samples. (Phyloseq does not allow subsetting by sample name, only by sample metadata.)


```{r, echo=FALSE}
dada<-read.table("dada_read_stats.txt",sep="\t",header=TRUE)
names(dada)[1]<-"sample"
meta<-read.table("metadata.txt",sep="\t",header=TRUE)
meta_dada<-merge(meta,dada,by="sample")
write.table(meta_dada, "metadata.txt",sep="\t",col.names=NA)
# I had to manually remove the first column of rownumbers before using in phyloseq
```


## Assign taxonomy in DADA2

Make sure the taxonomy reference database is in your working directory. Keep the database file gzipped. Adjust path name below. This step is very time consuming.

When taxonomy assignment is complete, we will use base R and phyloseq to clean up the taxonomy table. First, we will replace NAs and empty cells with the lowest taxonomy classification available. Second, we will use phyloseq to remove reads that are classified as Eukaryotes or unclassified at the domain level (ie, we are keeping only Bacteria and Archaea because that is what our primers target).

```{r, echo=FALSE}
taxa <- assignTaxonomy(seqtab.nochim, "~/Documents/silva_nr99_v138.2_toGenus_trainset.fa.gz", multithread=TRUE)
# FIX the NAs in the taxa table
taxon <- as.data.frame(taxa,stringsAsFactors=FALSE)
taxon$Phylum[is.na(taxon$Phylum)] <- taxon$Kingdom[is.na(taxon$Phylum)]
taxon$Class[is.na(taxon$Class)] <- taxon$Phylum[is.na(taxon$Class)]
taxon$Order[is.na(taxon$Order)] <- taxon$Class[is.na(taxon$Order)]
taxon$Family[is.na(taxon$Family)] <- taxon$Order[is.na(taxon$Family)]
taxon$Genus[is.na(taxon$Genus)] <- taxon$Family[is.na(taxon$Genus)]
write.table(taxon,"silva_taxa_table.txt",sep="\t",col.names=NA)
write.table(seqtab.nochim, "silva_otu_table.txt",sep="\t",col.names=NA)
```


## Load data into phyloseq

```{r, echo=FALSE}
# Create phyloseq object from otu and taxonomy tables from dada2, along with the sample metadata.
otu <- read.table("silva_otu_table.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_taxa_table.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps # 2313 taxa and 184 samples
```

## Remove chloroplasts, mitochondria, and eukaryotes

```{r, echo=FALSE}
# remove chloroplasts and mitochondria and Eukaryota
get_taxa_unique(ps, "Family") #321
get_taxa_unique(ps, "Order") #195
get_taxa_unique(ps, "Kingdom") #3
ps <- subset_taxa(ps, Family !="Mitochondria")
ps <- subset_taxa(ps, Order !="Chloroplast")
ps <- subset_taxa(ps, Kingdom !="Eukaryota")
ps <- subset_taxa(ps, Kingdom !="NA")
get_taxa_unique(ps, "Family") #318
get_taxa_unique(ps, "Order") #193
get_taxa_unique(ps, "Kingdom") #2
ps
# 2064 taxa and 184 samples
```

## Export tables for future reference

```{r, echo=FALSE}
# Now export cleaned otu and taxa tables from phyloseq for future reference
otu = as(otu_table(ps), "matrix")
taxon = as(tax_table(ps), "matrix")
metadata = as(sample_data(ps), "matrix")
write.table(otu,"silva_nochloronomito_otu_table.txt",sep="\t",col.names=NA)
write.table(taxon,"silva_nochloronomito_taxa_table.txt",sep="\t",col.names=NA)
```

## Explore data from nochloronomito tables

Now, time to explore the data. First, I will remove the blanks and save the phyloseq object and tables without the blanks. Then, I will remove low abundance reads.

```{r, echo=FALSE}
# read data in, if needed
otu <- read.table("silva_nochloronomito_otu_table.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))

ps # 2064 taxa and 184 samples

# Remove sample blanks
psnb = subset_samples(ps, colony != "blank")
psnb # 2064 taxa and 180 samples

# Remove experimental samples with fewer than 500 nonchimeric reads
ps_qf = subset_samples(psnb, nonchim > 500)
ps_qf # 2064 taxa and 162 samples; 18 samples removed

# remove any empty rows
ps_qf <- prune_taxa(taxa_sums(ps_qf) > 1, ps_qf) 
ps_qf # 2032 taxa and 162 samples, 32 taxa removed

# export ASVs in blanks
ps_blanks = subset_samples(ps, colony == "blank")
ps_blanks <- prune_taxa(taxa_sums(ps_blanks) > 1, ps_blanks)
ps_blanks #6 taxa and 4 samples
get_taxa_unique(ps_blanks, "Genus")
# [1] "Bradyrhizobium"   "Ralstonia"        "Cloacibacterium"  "Pseudomonas"      "Actinotalea"      "Methylobacterium"
otu_ps_blanks = as(otu_table(ps_blanks), "matrix")
taxon_ps_blanks = as(tax_table(ps_blanks), "matrix")
write.table(otu_ps_blanks,"silva_nochloronomito_otu_table_blanks.txt",sep="\t",col.names=NA)
write.table(taxon_ps_blanks,"silva_nochloronomito_taxa_table_blanks.txt",sep="\t",col.names=NA)

# export tables with blanks and samples with fewer than 500 reads removed
otu_ps = as(otu_table(ps_qf), "matrix")
taxon_ps = as(tax_table(ps_qf), "matrix")
metadata_ps = as(sample_data(ps_qf), "matrix")
write.table(otu_ps,"silva_nochloronomito_otu_table_qf.txt",sep="\t",col.names=NA)
write.table(taxon_ps,"silva_nochloronomito_taxa_table_qf.txt",sep="\t",col.names=NA)
write.table(metadata_ps,"metadata_qf.txt",sep="\t",col.names=NA)

# export relative abundances
ps_ra<-transform_sample_counts(ps_qf, function(OTU) OTU/sum(OTU))
otu_ps_ra = as(otu_table(ps_ra), "matrix")
write.table(otu_ps_ra,"silva_nochloronomito_otu_table_RA.txt",sep="\t",col.names=NA)

#explore data
ntaxa(ps_qf) # 2032
get_taxa_unique(ps_qf, "Order") # 191
get_taxa_unique(ps_qf, "Class") # 89

ps5<-filter_taxa(ps_qf, function(x) mean(x) >5, TRUE)
ntaxa(ps5) # 38

ps2<-filter_taxa(ps_qf, function(x) mean(x) >2, TRUE)
ps2 # 89 taxa and 162 samples
ntaxa(ps2) # 89 ASVs
get_taxa_unique(ps2, "Class") # 14 classes
get_taxa_unique(ps2, "Order") # 33 orders
get_taxa_unique(ps2, "Family") # 46 families
get_taxa_unique(ps2, "Genus") # 60 genera

# Export files with low abundance taxa removed
otu_ps2 = as(otu_table(ps2), "matrix")
taxon_ps2 = as(tax_table(ps2), "matrix")
metadata_ps2 = as(sample_data(ps2), "matrix")
write.table(otu_ps2,"silva_nochloronomito_otu_table_ps2.txt",sep="\t",col.names=NA)
write.table(taxon_ps2,"silva_nochloronomito_taxa_table_ps2.txt",sep="\t",col.names=NA)
write.table(metadata_ps2,"metadata_ps2.txt",sep="\t",col.names=NA)
# also write out ASV table with relative abundance
ps2_ra<-transform_sample_counts(ps2, function(OTU) OTU/sum(OTU))
otu_ps2_ra = as(otu_table(ps2_ra), "matrix")
write.table(otu_ps2_ra,"silva_nochloronomito_otu_table_ps2_RA.txt",sep="\t",col.names=NA)
```


## Stacked bar charts for first look at composition

```{r, echo=FALSE}
# load in data and create phyloseq object
otu <- read.table("silva_nochloronomito_otu_table_ps2_RA.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_ps2.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_ps2.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps2_ra <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps2_ra # 89 taxa and 152 samples
# removed 10 samples (7 CNAT, 3 DLAB that only have counts that were mitochondria or chloroplasts - ie. no bacterial reads) from otu table and metadata table, when you convert to relative abundance it puts "NA" into OTU table; samples show up on x-axis but have stacked bar
# CNAT samples removed: CN041-1, CN041-3, CN042-1, CN042-2, CN042-3, CN043-2, CN043-3
# DLAB samples removed: DL047-2, DL055-2, DL056-1

get_taxa_unique(ps2_ra, "Class") # 14 classes
get_taxa_unique(ps2_ra, "Order") # 33 orders
get_taxa_unique(ps2_ra, "Family") # 46 families
get_taxa_unique(ps2_ra, "Genus") # 60 genera

# Plot orders
# generate high-contrast color combinations
n <- 33
# after plotting, you can re-run the next line to create a different selection of colors
palette <- distinctColorPalette(n)

# subset coral species to generate separate plots
ps2_cnat = subset_samples(ps2_ra, coral.species == "Colpophyllia natans")
ps2_cnat  # 89 taxa and 75 samples
sample_data(ps2_cnat)$timepoint<-factor(sample_data(ps2_cnat)$timepoint,levels=c("Initial","2-month"))

ps2_dlab = subset_samples(ps2_ra, coral.species == "Diploria labyrinthiformis")
ps2_dlab  # 89 taxa and 77 samples
sample_data(ps2_dlab)$timepoint<-factor(sample_data(ps2_dlab)$timepoint,levels=c("Initial","2-month"))

# look at individual colonies
p1a=plot_bar(ps2_cnat, fill="Order")+
  facet_grid(.~timepoint,scales="free",space="free")+
  geom_bar(aes(fill=Order), stat="identity",position="stack")+
  theme_bw()+
  theme(strip.text=element_text(face="bold", size=12))+
  theme(axis.text.x=element_text(size=12,angle=90,vjust=0.5))+
  theme(axis.text.y=element_text(size=12))+
  scale_fill_manual(values=palette)+
  #theme(axis.title.x = element_blank())+
  theme(legend.position = "bottom")
p1a

# look at individual colonies
p2a=plot_bar(ps2_dlab, fill="Order")+
  facet_grid(.~timepoint,scales="free",space="free")+
  geom_bar(aes(fill=Order), stat="identity",position="stack")+
  theme_bw()+
  theme(strip.text=element_text(face="bold", size=12))+
  theme(axis.text.x=element_text(size=12,angle=90,vjust=0.5))+
  theme(axis.text.y=element_text(size=12))+
  scale_fill_manual(values=palette)+
  #theme(axis.title.x = element_blank())+
  theme(legend.position = "bottom")
p2a


# Plot families
get_taxa_unique(ps2_ra, "Family") # 46 families
# generate high-contrast color combinations
n <- 46
# after plotting, you can re-run the next line to create a different selection of colors
palette2 <- distinctColorPalette(n)

pdf("barchart_Family_CNAT.pdf",width=11)
p1a=plot_bar(ps2_cnat, fill="Family")+
  facet_grid(.~timepoint,scales="free",space="free")+
  geom_bar(aes(fill=Family), stat="identity",position="stack")+
  theme_bw()+
  theme(strip.text=element_text(face="bold", size=12))+
  theme(axis.text.x=element_text(size=10,angle=90,vjust=0.5))+
  theme(axis.text.y=element_text(size=12))+
  scale_fill_manual(values=palette2)+
  theme(axis.title.x = element_blank())+
  theme(legend.position = "bottom")
p1a
dev.off()

pdf("barchart_Family_DLAB.pdf",width=11)
p2a=plot_bar(ps2_dlab, fill="Family")+
  facet_grid(.~timepoint,scales="free",space="free")+
  geom_bar(aes(fill=Family), stat="identity",position="stack")+
  theme_bw()+
  theme(strip.text=element_text(face="bold", size=12))+
  theme(axis.text.x=element_text(size=10,angle=90,vjust=0.5))+
  theme(axis.text.y=element_text(size=12))+
  scale_fill_manual(values=palette2)+
  theme(axis.title.x = element_blank())+
  theme(legend.position = "bottom")
p2a
dev.off()


```


## Bubble Plot of most abundant bacterial families

```{r, echo=FALSE}
# load in data and create phyloseq object
otu <- read.table("silva_nochloronomito_otu_table_ps2_RA.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_ps2.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_ps2.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps2_ra <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps2_ra # 89 taxa and 152 samples
get_taxa_unique(ps2_ra, "Family") # 46 families

# Plotting most abundant families - first collapse ASVs into families then reshape data for plotting
ps2_ra_f = tax_glom(ps2_ra, "Family")
ps2_ra_f # 46 families and 152 samples
otu_f = as(otu_table(ps2_ra_f), "matrix")
taxon_f = as(tax_table(ps2_ra_f), "matrix")
fam<-t(otu_f)
fam <- as.data.frame(fam)
fam <- tibble::rownames_to_column(fam, "ASV")
taxon_f <- as.data.frame(taxon_f)
taxon_f <- tibble::rownames_to_column(taxon_f, "ASV")
fam_tax <- merge(fam, taxon_f, by="ASV", all=TRUE)
fam_tax$Genus<-NULL
fam_tax$ASV<-NULL
# Order by total relative abundance and choose top 25 families
fam_tax <- fam_tax %>% mutate("total" = rowSums((fam_tax[,1:152]), na.rm = TRUE))  
fam_tax <- fam_tax[order(fam_tax$total, decreasing = TRUE),]
fam25 <- fam_tax[1:25,]
fam25$total<-NULL
# calculate a row with the relative abundance of all other families not included in the top 25 (21 other families), first remove taxonomy
fam25_just25<-fam25[1:152]
# next add up the relative abundances of the top 25 families
fam25_25total<-rbind(fam25_just25, colSums(fam25_just25[,1:152]))
# calculate what is the relative abundance of the other families
fam25_others<-rbind(fam25_25total, 1-fam25_25total[26,])
# add "Others" as value for 5 taxonomy columns
fam25_others$Kingdom = "Others"
fam25_others$Phylum = "Others"
fam25_others$Class = "Others"
fam25_others$Order = "Others"
fam25_others$Family = "Others"
# remove all but the last row
others<-fam25_others[-(1:26),]
# finally, add only the last row back to the original dataframe fam25
fam25_withothers <- rbind(fam25, others)

fam_long<-melt(fam25_withothers,value.name="proportion",variable.name="sample",id.vars=c("Kingdom","Phylum","Class","Order","Family"))
# Merge with metadata for plotting aesthetics
samples <- as.data.frame(samples)
samples <- tibble::rownames_to_column(samples, "sample")
famplot <- merge(fam_long,samples,by="sample", all=TRUE)

# separate by coral species
famplotC <- famplot %>% filter(coral.species =='Colpophyllia natans')
unique(famplotC$Class) # 12 classes
# [1] "Gammaproteobacteria" "Bacteroidia"         "Alphaproteobacteria" "Blastocatellia"      "Campylobacteria"    
# [6] "Cyanobacteriia"      "Nitrososphaeria"     "Desulfovibrionia"    "Desulfobulbia"       "Clostridia"         
# [11] "Bacteria"            "Others"  

famplotD <- famplot %>% filter(coral.species =='Diploria labyrinthiformis')
unique(famplotD$Class) # same 12 classes as above

#set colorblind friendly palette, this is muted_nine plus black and white
cols<-c("Alphaproteobacteria"="#CC6677","Bacteria"="#117733","Bacteroidia"="#332288","Blastocatellia"="#88CCEE","Campylobacteria"="#999933","Clostridia"="#882255","Cyanobacteriia"="#44AA99","Desulfobulbia"="#DDCC77","Desulfovibrionia"="#AA4499","Gammaproteobacteria"="#000000","Nitrososphaeria"="#F9F4EC","Others"="#999999")


# set aesthetics for each coral species
famplotC$timepoint<-factor(famplotC$timepoint,levels=c("Initial","2-month"))
famplotC$Family<-factor(famplotC$Family,levels=c("Others","Bacteroidia","Burkholderiaceae","Rhodospirillaceae","Paraspirulinaceae","AB1","Alphaproteobacteria","Kiloniellaceae","Bacteria","Fusibacteraceae","Woeseiaceae","Endozoicomonadaceae","Desulfocapsaceae","Desulfovibrionaceae","Nitrosopumilaceae","Cyanobacteriaceae","Phormidiaceae","Campylobacterales","Hyphomonadaceae","Flavobacteriaceae","Blastocatellaceae","Paracoccaceae","Terasakiellaceae","Vibrionaceae","Amoebophilaceae","Ga0077536"))

#order samples by site West to East (lower keys = Looe, middle keys = West Turtle, upper keys = Pickles)
famplotC$sample<-factor(famplotC$sample,levels=c("CN034-1","CN034-2","CN034-3","CN034-4","CN034-5","CN034-6","CN035-1","CN035-2","CN035-3","CN035-4","CN035-5","CN035-6","CN036-3","CN036-4","CN036-5","CN036-6","CN037-1","CN037-2","CN037-3","CN037-4","CN037-5","CN037-6","CN038-1","CN038-2","CN038-3","CN038-4","CN038-5","CN038-6","CN021-1","CN021-2","CN021-3","CN021-4","CN021-5","CN021-6","CN022-1","CN022-2","CN022-4","CN022-5","CN022-6","CN026-1","CN026-3","CN026-4","CN026-5","CN026-6","CN027-1","CN027-2","CN027-3","CN027-4","CN027-5","CN027-6","CN028-1","CN028-2","CN028-3","CN028-4","CN028-5","CN028-6","CN041-2","CN041-4","CN041-5","CN041-6","CN042-4","CN042-5","CN042-6","CN043-1","CN043-4","CN043-5","CN043-6","CN044-1","CN044-4","CN044-5","CN044-6","CN045-1","CN045-4","CN045-5","CN045-6"))
unique(famplotC$sample)

famplotD$timepoint<-factor(famplotD$timepoint,levels=c("Initial","2-month"))
famplotD$Family<-factor(famplotD$Family,levels=c("Others","Bacteroidia","Burkholderiaceae","Rhodospirillaceae","Paraspirulinaceae","AB1","Alphaproteobacteria","Kiloniellaceae","Bacteria","Fusibacteraceae","Woeseiaceae","Endozoicomonadaceae","Desulfocapsaceae","Desulfovibrionaceae","Nitrosopumilaceae","Cyanobacteriaceae","Phormidiaceae","Campylobacterales","Hyphomonadaceae","Flavobacteriaceae","Blastocatellaceae","Paracoccaceae","Terasakiellaceae","Vibrionaceae","Amoebophilaceae","Ga0077536"))

pdf("Top25families_CNAT.pdf", bg ="white", width=11)
pc <- ggplot(famplotC, aes(sample, Family, fill= Class))+
  geom_point(aes(size=proportion*100),alpha=0.8,shape=21)+
  scale_size(range = c(0,10), name="Relative abundance")+
  scale_fill_manual(values = cols)+
  facet_grid(.~timepoint,scales="free",space="free")+
  theme_bw()+
  theme(axis.text.x=element_text(angle=90))+
  theme(axis.title.x=element_blank())+
  theme(axis.title.y=element_blank())
pc
dev.off()

pdf("Top25families_DLAB.pdf", bg ="white", width=11)
pd <- ggplot(famplotD, aes(sample, Family, fill= Class))+
  geom_point(aes(size=proportion*100),alpha=0.8,shape=21)+
  scale_size(range = c(0,10), name="Relative abundance")+
  scale_fill_manual(values = cols)+
  facet_grid(.~timepoint,scales="free",space="free")+
  theme_bw()+
  theme(axis.text.x=element_text(angle=90))+
  theme(axis.title.x=element_blank())+
  theme(axis.title.y=element_blank())
pd
dev.off()

```


# Beta diversity


```{r, echo=FALSE}
# load in data and create phyloseq object
otu <- read.table("silva_nochloronomito_otu_table_qf.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_qf.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_qf.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps_qf <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps_qf #2032 taxa and 152 samples
# removed 10 samples as above for ps2 that only have counts that were mitochondria or chloroplasts 

# First, replace 0 values with an estimate (because normalization is taking log, can't have 0)
# Also transposing here, need samples as rows
d.czm <- cmultRepl(t(otu), method="CZM", label=0, z.warning=1)
# Perform the center-log-ratio (CLR) transformation 
d.clr <- codaSeq.clr(d.czm)
# transpose matrix of CLR transformed data for ordination and dendrogram
E.clr <- t(d.clr)
# plot compositional PCA biplot (perform a singular value decomposition)
d.pcx <- prcomp(E.clr)
# calculate percent variance explained for the axis labels
pc1 <- round(d.pcx$sdev[1]^2/sum(d.pcx$sdev^2),2)
pc2 <- round(d.pcx$sdev[2]^2/sum(d.pcx$sdev^2),2)
xlab <- paste("PC1: ", pc1, sep="")
ylab <- paste("PC2: ", pc2, sep="")
biplot(d.pcx, cex=c(0.6,0.4), var.axes=F,scale=1, xlab=xlab, ylab=ylab)
summary(d.pcx)
str(d.pcx)
screeplot(d.pcx)
# Make a pretty PCA plot with ggplot
df_out <- as.data.frame(d.pcx$x)
theme_set(theme_bw()+theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank()))
#cols<-c("Colpophyllia natans"="#56B4E9","Diploria labyrinthiformis"="#0072B2")
cols<-c("Initial"="#56B4E9","2-month"="#0072B2")
samples$timepoint<-factor(samples$timepoint,levels=c("Initial","2-month"))
samples$site<-factor(samples$site,levels=c("Looe Key Inshore","West Turtle","Pickles Patch"))

pdf("PCA.pdf",bg ="white",width=11)
p<-ggplot(df_out,aes(x=PC1,y=PC2,color=samples$timepoint,shape=samples$site))
p<-p+geom_point(size=3)+
  theme(axis.title = element_text(size=14))+
  theme(axis.text=element_text(size=14))+
  theme(legend.title = element_text(size=14))+
  theme(legend.text = element_text(size=14))+
  theme(strip.text.x = element_text(size=14, face="italic"))+
  scale_color_manual(values=cols)+
  guides(fill = guide_legend(override.aes=list(shape=21)))+
  labs(x=xlab, y=ylab, color="Timepoint",shape="Site")+
  facet_grid(.~samples$coral.species)+
  ylim(-50,50)+
  xlim(-90,90)+
  coord_fixed(ratio=1)
p
dev.off()

####### Use phyloseq/vegan to perform PERMANOVA
# set metadata as factors
time<-as.character(samples$timepoint)
site<-as.character(samples$site)
species<-as.character(samples$coral.species)
colony<-as.character(samples$colony)

# permanova between groups using Aitchison distance
dist.clr <- dist(E.clr)

perm<-adonis2(dist.clr~coral.species*timepoint,as(sample_data(ps_qf),"data.frame"))
print(perm)
# coral species, time, and the interaction of species and time are significant at 0.05 level, but all the R2 values are very low

perm1<-adonis2(dist.clr~colony*timepoint,as(sample_data(ps_qf),"data.frame"))
print(perm1) 
# interaction of colony and time is significant at 0.05 level, R2 = 0.22632

perm1a<-adonis2(dist.clr~colony,as(sample_data(ps_qf),"data.frame"))
print(perm1a) # colony: not significant

perm2<-pairwise.adonis2(dist.clr~site/coral.species,as(sample_data(ps_qf),"data.frame"), strata='coral.species')
print(perm2)
# interaction of site and coral species significant only between LK and Pickles at 0.05 level

perm3<-pairwise.adonis2(dist.clr~timepoint/colony,as(sample_data(ps_qf),"data.frame"), strata='colony')
print(perm3)
# time, controlling for colony is significant at 0.05 level, R2 = 0.01422; interaction of time and colony sign. at 0.05 level, R2 = 0.43672

perm4<-pairwise.adonis2(dist.clr~site/coral.species,as(sample_data(ps_qf),"data.frame"), strata='coral.species')
print(perm4) # site, controlling for coral species = not significant

perm5<-pairwise.adonis2(dist.clr~time/coral.species,as(sample_data(ps_qf),"data.frame"), strata='coral.species')
print(perm5) # time, controlling for coral species is significant at 0.05 level, R2 = 0.01422; interaction of time and coral species sign. at 0.05 level, R2 = 0.03216

```


## Look at permanova for only the 2-month samples because it has all 90 samples


```{r, echo=FALSE}
# load in data and create phyloseq object
otu <- read.table("silva_nochloronomito_otu_table_qf.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_qf.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_qf.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps_qf <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps_qf #2032 taxa and 152 samples

ps_2mon = subset_samples(ps_qf, timepoint == "2-month")
ps_2mon #2032 taxa and 90 samples
#export the tables from only 2-month samples
otu_2 = as(otu_table(ps_2mon), "matrix")
taxon_2 = as(tax_table(ps_2mon), "matrix")
metadata_2 = as(sample_data(ps_2mon), "matrix")
write.table(otu_2,"silva_nochloronomito_otu_table_2month.txt",sep="\t",col.names=NA)
write.table(taxon_2,"silva_nochloronomito_taxa_table_2month.txt",sep="\t",col.names=NA)
write.table(metadata_2,"metadata_2month.txt",sep="\t",col.names=NA)

# CLEAR OUT ENVIRONMENT and read back in tables because the clr is done on the otu table (not on the phyloseq object)
otu <- read.table("silva_nochloronomito_otu_table_2month.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_2month.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_2month.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps # 2032 taxa and 90 samples

# First, replace 0 values with an estimate (because normalization is taking log, can't have 0)
# Also transposing here, need samples as rows
d.czm <- cmultRepl(t(otu), method="CZM", label=0, z.warning=1)
# Perform the center-log-ratio (CLR) transformation 
d.clr <- codaSeq.clr(d.czm)
# transpose matrix of CLR transformed data for ordination and dendrogram
E.clr <- t(d.clr)
# plot compositional PCA biplot (perform a singular value decomposition)
d.pcx <- prcomp(E.clr)
# calculate percent variance explained for the axis labels
pc1 <- round(d.pcx$sdev[1]^2/sum(d.pcx$sdev^2),2)
pc2 <- round(d.pcx$sdev[2]^2/sum(d.pcx$sdev^2),2)
xlab <- paste("PC1: ", pc1, sep="")
ylab <- paste("PC2: ", pc2, sep="")
biplot(d.pcx, cex=c(0.6,0.4), var.axes=F,scale=1, xlab=xlab, ylab=ylab)
summary(d.pcx)
str(d.pcx)
screeplot(d.pcx)
# Make a pretty PCA plot with ggplot
df_out <- as.data.frame(d.pcx$x)
theme_set(theme_bw()+theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank()))
cols<-c("Colpophyllia natans"="#56B4E9","Diploria labyrinthiformis"="#0072B2")

pdf("PCA_2month.pdf",bg ="white",width=11)
p<-ggplot(df_out,aes(x=PC1,y=PC2,color=samples$coral.species,shape=samples$site))
p<-p+geom_point(size=3)+
  theme(axis.title = element_text(size=14))+
  theme(axis.text=element_text(size=14))+
  theme(legend.title = element_text(size=14))+
  theme(legend.text = element_text(size=14))+
  theme(strip.text.x = element_text(size=14, face="italic"))+
  scale_color_manual(values=cols)+
  guides(fill = guide_legend(override.aes=list(shape=21)))+
  labs(x=xlab, y=ylab, color="Coral species",shape="Site")+
  facet_grid(.~samples$coral.species)+
  ylim(-50,50)+
  xlim(-50,50)+
  coord_fixed(ratio=1)
p
dev.off()

####### Use phyloseq/vegan to perform PERMANOVA
# set metadata as factors
site<-as.character(samples$site)
species<-as.character(samples$coral.species)
colony<-as.character(samples$colony)
# permanova between groups using Aitchison distance
dist.clr <- dist(E.clr)

perm<-pairwise.adonis2(dist.clr~site/coral.species,as(sample_data(ps),"data.frame"), strata='coral.species')
print(perm) # site, controlling for coral species = not significant

perm2<-adonis2(dist.clr~colony,as(sample_data(ps),"data.frame"))
print(perm2) # colony (genotype) = not significant

perm3<-adonis2(dist.clr~species,as(sample_data(ps),"data.frame"))
print(perm3) # species = not significant

```

# Differential abundance of taxa over time: CNAT only

```{r, echo=FALSE}
otu <- read.table("silva_nochloronomito_otu_table_qf.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_qf.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_qf.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps_qf <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps_qf #2032 taxa and 152 samples

#Select CNAT samples only
ps_cnat = subset_samples(ps_qf, coral.species == "Colpophyllia natans")
ps_cnat #2032 taxa and 75 samples

# collapse ASVs into families
ps_fam = tax_glom(ps_cnat, "Family")
ps_fam # 320 families and 75 samples
ps_fam <- prune_taxa(taxa_sums(ps_fam) > 1, ps_fam) #remove taxa not found in remaining samples
ps_fam #244 families and 75 samples

# DA determined relative to first reference level, so set this manually to the initial timepoint
sample_data(ps_fam)$timepoint<-factor(sample_data(ps_fam)$timepoint,levels=c("Initial","2-month"))

set.seed(1)
time.da <- differentialTest(formula = ~ timepoint, 
                                 phi.formula = ~ timepoint,
                                 formula_null = ~ 1,
                                 phi.formula_null = ~ timepoint,
                                 test = "Wald", boot = FALSE,
                                 data = ps_fam,
                                 fdr_cutoff = 0.05)

summary(time.da)
time.da
time.da$significant_taxa #1

#quick look at results with corncob's plotting feature
plot(time.da, c("Family"))
# Positive values mean that taxon is enriched at 2-month timepoint relative to initial samples
# CNAT: only Blastocatellaceae is DA over time, more abundant in initial timepoint 

# create a function that will return the values of interest
sigtaxto_df <- function(treatment.da, ps_fam) {
  df.out <- c()
  for (i in 1:length(treatment.da$significant_models)) { #from 1 to the number of significant taxa
    df.out = rbind(df.out, treatment.da$significant_models[[i]]$coefficients[2, 1:4]) #bind the significant taxa with coefficient, std. error, t value, Pr. coefficients[2,] is the DA model output.
  }
  asv = as.vector(treatment.da$significant_taxa) #get significant families
  Class = as.vector(otu_to_taxonomy(data = ps_fam, time.da$significant_taxa, level = c("Class")))
  Order = as.vector(otu_to_taxonomy(data = ps_fam, time.da$significant_taxa, level = c("Order")))
  Family = as.vector(otu_to_taxonomy(data = ps_fam, time.da$significant_taxa, level = c("Family")))
  
  df.bind <- as.data.frame(cbind(df.out, asv, Class, Order, Family))
  colnames(df.bind) <- c("Estimate", "StdError", "tvalue", "Pr", "asv", "Class", "Order", "Family")
  df.bind$Estimate = as.numeric(as.character(df.bind$Estimate)) #estimate column should be numeric 
  df.bind$StdError = as.numeric(as.character(df.bind$StdError)) #stderror column should be numeric 
  
  return(df.bind)
}

df <- sigtaxto_df(time.da, ps_fam) #make a df of the corncob output and the taxonomy
### Save these data so you don't have to re-run the model ###
write.table(df, "DA-Families_time_CNAT.txt", sep="\t",row.names = FALSE)


```


# Differential abundance of taxa over time: DLAB only

```{r, echo=FALSE}
otu <- read.table("silva_nochloronomito_otu_table_qf.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_qf.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_qf.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps_qf <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps_qf #2032 taxa and 152 samples

#Select DLAB samples only
ps_dlab = subset_samples(ps_qf, coral.species == "Diploria labyrinthiformis")
ps_dlab #2032 taxa and 77 samples

# collapse ASVs into families
ps_fam = tax_glom(ps_dlab, "Family")
ps_fam # 320 families and 77 samples
ps_fam <- prune_taxa(taxa_sums(ps_fam) > 1, ps_fam) #remove taxa not found in remaining samples
ps_fam #255 families and 77 samples

# DA determined relative to first reference level, so set this manually to the initial timepoint
sample_data(ps_fam)$timepoint<-factor(sample_data(ps_fam)$timepoint,levels=c("Initial","2-month"))

set.seed(1)
time.da <- differentialTest(formula = ~ timepoint, 
                                 phi.formula = ~ timepoint,
                                 formula_null = ~ 1,
                                 phi.formula_null = ~ timepoint,
                                 test = "Wald", boot = FALSE,
                                 data = ps_fam,
                                 fdr_cutoff = 0.05)

summary(time.da)
time.da
time.da$significant_taxa #4

#quick look at results with corncob's plotting feature
plot(time.da, c("Family"))
# Positive values mean that taxon is enriched at 2-month timepoint relative to initial samples
# DLAB: Nitrosopumilaceae and Caldilineaceae more abundant in initial timepoint, Vibrionaceae and Campylobacterales more abundant at 2 months

# create a function that will return the values of interest
sigtaxto_df <- function(treatment.da, ps_fam) {
  df.out <- c()
  for (i in 1:length(treatment.da$significant_models)) { #from 1 to the number of significant taxa
    df.out = rbind(df.out, treatment.da$significant_models[[i]]$coefficients[2, 1:4]) #bind the significant taxa with coefficient, std. error, t value, Pr. coefficients[2,] is the DA model output.
  }
  asv = as.vector(treatment.da$significant_taxa) #get significant families
  Class = as.vector(otu_to_taxonomy(data = ps_fam, time.da$significant_taxa, level = c("Class")))
  Order = as.vector(otu_to_taxonomy(data = ps_fam, time.da$significant_taxa, level = c("Order")))
  Family = as.vector(otu_to_taxonomy(data = ps_fam, time.da$significant_taxa, level = c("Family")))
  
  df.bind <- as.data.frame(cbind(df.out, asv, Class, Order, Family))
  colnames(df.bind) <- c("Estimate", "StdError", "tvalue", "Pr", "asv", "Class", "Order", "Family")
  df.bind$Estimate = as.numeric(as.character(df.bind$Estimate)) #estimate column should be numeric 
  df.bind$StdError = as.numeric(as.character(df.bind$StdError)) #stderror column should be numeric 
  
  return(df.bind)
}

df <- sigtaxto_df(time.da, ps_fam) #make a df of the corncob output and the taxonomy
### Save these data so you don't have to re-run the model ###
write.table(df, "DA-Families_time_DLAB.txt", sep="\t",row.names = FALSE)

```


# Differential abundance of taxa over time, controlling for coral species

```{r, echo=FALSE}
otu <- read.table("silva_nochloronomito_otu_table_qf.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_qf.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_qf.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps_qf <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps_qf #2032 taxa and 152 samples

# collapse ASVs into families
ps_fam = tax_glom(ps_qf, "Family")
ps_fam # 320 families and 152 samples
ps_fam <- prune_taxa(taxa_sums(ps_fam) > 1, ps_fam) #remove taxa not found in remaining samples
ps_fam #318 families and 152 samples

# DA determined relative to first reference level, so set this manually to the initial timepoint
sample_data(ps_fam)$timepoint<-factor(sample_data(ps_fam)$timepoint,levels=c("Initial","2-month"))

set.seed(1)
time.da <- differentialTest(formula = ~ timepoint + coral.species, 
                                 phi.formula = ~ timepoint,
                                 formula_null = ~ coral.species,
                                 phi.formula_null = ~ 1,
                                 test = "Wald", boot = FALSE,
                                 data = ps_fam,
                                 fdr_cutoff = 0.05)

summary(time.da)
time.da
time.da$significant_taxa #18

#quick look at results with corncob's plotting feature
plot(time.da, c("Family"))
# Positive values mean that taxon is enriched at 2-month timepoint relative to initial samples

# create a function that will return the values of interest
sigtaxto_df <- function(time.da, ps_fam) {
  df.out <- c()
  for (i in 1:length(time.da$significant_models)) { #from 1 to the number of significant taxa
    df.out = rbind(df.out, time.da$significant_models[[i]]$coefficients[2, 1:4]) #bind the significant taxa with coefficient, std. error, t value, Pr. coefficients[2,] is the DA model output.
  }
  asv = as.vector(time.da$significant_taxa) #get significant families
  Class = as.vector(otu_to_taxonomy(data = ps_fam, time.da$significant_taxa, level = c("Class")))
  Order = as.vector(otu_to_taxonomy(data = ps_fam, time.da$significant_taxa, level = c("Order")))
  Family = as.vector(otu_to_taxonomy(data = ps_fam, time.da$significant_taxa, level = c("Family")))
  
  df.bind <- as.data.frame(cbind(df.out, asv, Class, Order, Family))
  colnames(df.bind) <- c("Estimate", "StdError", "tvalue", "Pr", "asv", "Class", "Order", "Family")
  df.bind$Estimate = as.numeric(as.character(df.bind$Estimate)) #estimate column should be numeric 
  df.bind$StdError = as.numeric(as.character(df.bind$StdError)) #stderror column should be numeric 
  
  return(df.bind)
}

df <- sigtaxto_df(time.da, ps_fam) #make a df of the corncob output and the taxonomy
### Save these data so you don't have to re-run the model ###
write.table(df, "DA-Families_time_controlling-coral-species.txt", sep="\t",row.names = FALSE)


# families enriched at 2 months: Rickettsiaceae, Eurycoccales Incertae Sedis, Desulfovibrionaceae, Hyphomonadaceae, Vibrionaceae, Cellvibrionaceae, Terasakiellaceae, Campylobacterales

# families enriched at initial timepoint: Cyclobacteriaceae, Amoebophilaceae, Blastocatellaceae, Caminicellaceae, Microtrichaceae, Vicinamibacteria Subgroup 9, Vicinamibacterales, Caldilineaceae, Cyanobacteriaceae, Nitrosopumilaceae


```



# Differential abundance of taxa by site, controlling for coral species. Using 2-month samples only

```{r, echo=FALSE}
otu <- read.table("silva_nochloronomito_otu_table_2month.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_2month.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_2month.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps # 2032 taxa and 90 samples


# collapse ASVs into families
ps_fam = tax_glom(ps, "Family")
ps_fam # 320 families and 90 samples
ps_fam <- prune_taxa(taxa_sums(ps_fam) > 1, ps_fam) #remove taxa not found in remaining samples
ps_fam # 222 families and 90 samples

# DA determined relative to first reference level, so set this manually to Pickles Patch, the site that was most different environment (not influenced by heated, supersaline bay water)
sample_data(ps_fam)$site<-factor(sample_data(ps_fam)$site,levels=c("Pickles Patch","Looe Key Inshore","West Turtle"))

set.seed(1)
site.da <- differentialTest(formula = ~ site + coral.species, 
                                 phi.formula = ~ site,
                                 formula_null = ~ coral.species,
                                 phi.formula_null = ~ 1,
                                 test = "Wald", boot = FALSE,
                                 data = ps_fam,
                                 fdr_cutoff = 0.05)

summary(site.da)
site.da
site.da$significant_taxa #1

#quick look at results with corncob's plotting feature
plot(site.da, c("Family"))
# Positive values mean that taxon is enriched at the site relative to Pickles; both Looe Key and West Turtle are enriched in the unclassified family from the Gammaproteobacterial order Ga0077536

# create a function that will return the values of interest
sigtaxto_df <- function(site.da, ps_fam) {
  df.out <- c()
  for (i in 1:length(site.da$significant_models)) { #from 1 to the number of significant taxa
    df.out = rbind(df.out, site.da$significant_models[[i]]$coefficients[2, 1:4]) #bind the significant taxa with coefficient, std. error, t value, Pr. coefficients[2,] is the DA model output.
  }
  asv = as.vector(site.da$significant_taxa) #get significant families
  Class = as.vector(otu_to_taxonomy(data = ps_fam, site.da$significant_taxa, level = c("Class")))
  Order = as.vector(otu_to_taxonomy(data = ps_fam, site.da$significant_taxa, level = c("Order")))
  Family = as.vector(otu_to_taxonomy(data = ps_fam, site.da$significant_taxa, level = c("Family")))
  
  df.bind <- as.data.frame(cbind(df.out, asv, Class, Order, Family))
  colnames(df.bind) <- c("Estimate", "StdError", "tvalue", "Pr", "asv", "Class", "Order", "Family")
  df.bind$Estimate = as.numeric(as.character(df.bind$Estimate)) #estimate column should be numeric 
  df.bind$StdError = as.numeric(as.character(df.bind$StdError)) #stderror column should be numeric 
  
  return(df.bind)
}

df <- sigtaxto_df(site.da, ps_fam) #make a df of the corncob output and the taxonomy
### Save these data so you don't have to re-run the model ###
write.table(df, "DA-Families_site_controlling-coral-species.txt", sep="\t",row.names = FALSE)



```

# Create a bubble plot with DA families in the top 25


```{r, echo=FALSE}
otu <- read.table("silva_nochloronomito_otu_table_qf.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("silva_nochloronomito_taxa_table_qf.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("metadata_qf.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps_qf <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps_qf #2032 taxa and 152 samples

# collapse ASVs into families
ps_fam = tax_glom(ps_qf, "Family")
ps_fam # 320 families and 152 samples
ps_fam <- prune_taxa(taxa_sums(ps_fam) > 1, ps_fam) #remove taxa not found in remaining samples
ps_fam # 318 families and 152 samples

ps_fam_ra<-transform_sample_counts(ps_fam, function(OTU) OTU/sum(OTU))
ps_fam_ra # 318 families and 152 samples
otu_f = as(otu_table(ps_fam_ra), "matrix")
taxon_f = as(tax_table(ps_fam_ra), "matrix")
fam<-t(otu_f)
fam <- as.data.frame(fam)
fam <- tibble::rownames_to_column(fam, "ASV")
taxon_f <- as.data.frame(taxon_f)
taxon_f <- tibble::rownames_to_column(taxon_f, "ASV")
fam_tax <- merge(fam, taxon_f, by="ASV", all=FALSE)
fam_tax$Genus<-NULL
fam_tax$ASV<-NULL
fam_tax$sequence<-NULL
# now to retrieve only the 10 families out of 318 families that I want to plot: choosing the DA families that are also in the top 25 most abundant families
DA1 <- fam_tax %>% filter(Family =='Nitrosopumilaceae') # enriched in initial timepoint
DA2 <- fam_tax %>% filter(Family =='Cyanobacteriaceae') # enriched in initial timepoint (DLAB)
DA3 <- fam_tax %>% filter(Family =='Blastocatellaceae') # enriched in initial timepoint (CNAT)
DA4 <- fam_tax %>% filter(Family =='Amoebophilaceae') # enriched in initial timepoint
DA5 <- fam_tax %>% filter(Family =='Desulfovibrionaceae') # enriched at 2-month timpoint
DA6 <- fam_tax %>% filter(Family =='Hyphomonadaceae') # enriched at 2-month timpoint
DA7 <- fam_tax %>% filter(Family =='Vibrionaceae') # enriched at 2-month timpoint
DA8 <- fam_tax %>% filter(Family =='Terasakiellaceae') # enriched at 2-month timpoint
DA9 <- fam_tax %>% filter(Family =='Campylobacterales') # enriched at 2-month timpoint
DA10 <- fam_tax %>% filter(Family == 'Ga0077536') # depleted at Pickles Patch / enriched at Looe Key and West Turtle (esp. CNAT)
DA_nursery<-rbind(DA1,DA2,DA3,DA4,DA5,DA6,DA7,DA8,DA9,DA10)

# Merge with metadata for plotting aesthetics
samples <- as.data.frame(samples)
samples <- tibble::rownames_to_column(samples, "sample")
DA_nursery_long<-melt(DA_nursery,value.name="proportion",variable.name="sample",id.vars=c("Kingdom","Phylum","Class","Order","Family"))
DAplot <- merge(DA_nursery_long,samples,by="sample", all=TRUE)
unique(DAplot$Class) # 8 classes
# [1] "Nitrososphaeria"     "Cyanobacteriia"      "Blastocatellia"      "Bacteroidia"         "Desulfovibrionia"   
# [6] "Alphaproteobacteria" "Gammaproteobacteria" "Campylobacteria" 

# separate by coral species
DAplotC <- DAplot %>% filter(coral.species =='Colpophyllia natans')
unique(DAplotC$Class)  

DAplotD <- DAplot %>% filter(coral.species =='Diploria labyrinthiformis')
unique(DAplotD$Class)

#set colorblind friendly palette, this is muted_nine plus black and white
cols<-c("Alphaproteobacteria"="#CC6677","Bacteroidia"="#332288","Blastocatellia"="#88CCEE","Campylobacteria"="#999933","Cyanobacteriia"="#44AA99","Desulfovibrionia"="#AA4499","Gammaproteobacteria"="#000000","Nitrososphaeria"="#F9F4EC")

# set aesthetics for each coral species
DAplotC$timepoint<-factor(DAplotC$timepoint,levels=c("Initial","2-month"))
DAplotC$Family<-factor(DAplotC$Family,levels=c("Desulfovibrionaceae","Nitrosopumilaceae","Cyanobacteriaceae","Campylobacterales","Hyphomonadaceae","Blastocatellaceae","Terasakiellaceae","Vibrionaceae","Amoebophilaceae","Ga0077536"))

# set aesthetics for each coral species
DAplotD$timepoint<-factor(DAplotD$timepoint,levels=c("Initial","2-month"))
DAplotD$Family<-factor(DAplotD$Family,levels=c("Desulfovibrionaceae","Nitrosopumilaceae","Cyanobacteriaceae","Campylobacterales","Hyphomonadaceae","Blastocatellaceae","Terasakiellaceae","Vibrionaceae","Amoebophilaceae","Ga0077536"))

pdf("DA_families_CNAT.pdf", bg ="white", width=11)
pc <- ggplot(DAplotC, aes(sample, Family, fill= Class))+
  geom_point(aes(size=proportion*100),alpha=0.8,shape=21)+
  scale_size(range = c(0,10), name="Relative abundance")+
  scale_fill_manual(values = cols)+
  facet_grid(.~timepoint,scales="free",space="free")+
  theme_bw()+
  theme(axis.text.x=element_text(angle=90))+
  theme(axis.title.x=element_blank())+
  theme(axis.title.y=element_blank())
pc
dev.off()

pdf("DA_families_DLAB.pdf", bg ="white", width=11)
pd <- ggplot(DAplotD, aes(sample, Family, fill= Class))+
  geom_point(aes(size=proportion*100),alpha=0.8,shape=21)+
  scale_size(range = c(0,10), name="Relative abundance")+
  scale_fill_manual(values = cols)+
  facet_grid(.~timepoint,scales="free",space="free")+
  theme_bw()+
  theme(axis.text.x=element_text(angle=90))+
  theme(axis.title.x=element_blank())+
  theme(axis.title.y=element_blank())
pd
dev.off()

```

